core/src/main/java/hivemall/utils/math/MatrixUtils.java

```diff
@@ -493,4 +498,202 @@ public static RealMatrix solve(@Nonnull final RealMatrix L, @Nonnull final RealM
+        double s = Av.getNorm();
+
+        for (int i = 0, n = u.length; i < n; i++) {
+            u[i] = Av.getEntry(i) / s;
+        }
+
+        return s;
+    }
+
+    /**
+     * Lanczos tridiagonalization for a symmetric matrix C to make s * s tridiagonal matrix T.
+     *
+     * http://www.cas.mcmaster.ca/~qiao/publications/spie05.pdf
+     *
+     * @param C target symmetric matrix
+     * @param a initial vector
+     * @param T result is stored here
+     */
+    @Nonnull
+    public static void lanczosTridiagonalization(@Nonnull final RealMatrix C, @Nonnull final double[] a,
+            @Nonnull final RealMatrix T) {
+        Preconditions.checkArgument(Arrays.deepEquals(C.getData(), C.transpose().getData()),
+                "Target matrix C must be a symmetric matrix");
+        Preconditions.checkArgument(C.getColumnDimension() == a.length,
+                "Column size of A and length of a should be same");
+        Preconditions.checkArgument(T.getRowDimension() == T.getColumnDimension(),
+                "T must be a square matrix");
+
+        int s = T.getRowDimension();
+
+        // initialize T with zeros
+        T.setSubMatrix(new double[s][s], 0, 0);
+
+        RealVector a0 = new ArrayRealVector(new double[a.length]);
+        RealVector r = new ArrayRealVector(a);
+
+        double beta0 = 1.d;
+
+        for (int i = 0; i < s; i++) {
+            RealVector a1 = r.mapDivide(beta0);
+            RealVector Ca1 = C.operate(a1);
+
+            double alpha1 = a1.dotProduct(Ca1);
+
+            r = Ca1.add(a1.mapMultiply(-1.d * alpha1)).add(a0.mapMultiply(-1.d * beta0));
+
+            double beta1 = r.getNorm();
+
+            T.setEntry(i, i, alpha1);
+            if (i - 1 >= 0) {
+                T.setEntry(i, i - 1, beta0);
+            }
+            if (i + 1 < s) {
+                T.setEntry(i, i + 1, beta1);
+            }
+
+            a0 = a1.copy();
+            beta0 = beta1;
+        }
+    }
+
+    /**
+     * QR decomposition for a tridiagonal matrix T.
+     *
+     * https://gist.github.com/lightcatcher/8118181
+     * http://www.ericmart.in/blog/optimizing_julia_tridiag_qr
+     *
+     * @param T target tridiagonal matrix
+     * @param R output matrix for R which is the same shape as T
+     * @param Qt output matrix for Q.T which is the same shape an T
+     */
+    @Nonnull
+    public static void tridiagonalQR(@Nonnull final RealMatrix T,
+            @Nonnull final RealMatrix R, @Nonnull final RealMatrix Qt) {
+        int n = T.getRowDimension();
+        Preconditions.checkArgument(n == R.getRowDimension() && n == R.getColumnDimension(),
+                "T and R must be the same shape");
+        Preconditions.checkArgument(n == Qt.getRowDimension() && n == Qt.getColumnDimension(),
+                "T and Qt must be the same shape");
+
+        // initial R = T
+        R.setSubMatrix(T.getData(), 0, 0);
+
+        // initial Qt = identity
+        Qt.setSubMatrix(new double[n][n], 0, 0);
```

Comment:
create `static double[][] eye(n)` or something that returns a 2-D array with ones on the diagonal and zeros elsewhere.
https://docs.scipy.org/doc/numpy-1.10.1/reference/generated/numpy.eye.html